server/src/lib/artefacts.ts

```diff
@@ -0,0 +1,216 @@
+import { existsSync, mkdirSync } from 'fs';
+import { readFile, writeFile, unlink } from 'fs/promises';
+import * as path from 'path';
+import * as zlib from 'zlib';
+
+import { S3 } from '@aws-sdk/client-s3';
+import { config } from 'dotenv';
+
+config();
+
+const { PREFERRED_STORAGE_TYPE, AWS_S3_BUCKET_NAME } = process.env;
+const REGION = 'eu-central-1';
+
+enum StorageType {
+    ObjectStorage = 'S3',
+    FileStorage = 'FS'
+}
+
+// no explicit auth because it implicitly relies on env vars to exist
+// AWS_ACCESS_KEY_ID and AWS_SECRET_ACCESS_KEY
+const S3Client = new S3({
+    region: REGION,
+});
+
+// helper method to convert stream to buffer (S3 responds with streams)
+const streamToBuffer = (stream: any): Promise<Buffer> => new Promise(
+    (resolve, reject) => {
+        const chunks: Uint8Array[] = [];
+        stream.on('data', (chunk: Uint8Array) => chunks.push(chunk));
+        stream.on('error', reject);
+        stream.on('end', () => resolve(Buffer.concat(chunks)));
+    },
+);
+
+const compress = (input: string|Buffer): Buffer => zlib.gzipSync(input);
+const decompress = (input: Buffer): Buffer => zlib.unzipSync(input);
+
+// generic helper method for uploading artefacts (includes compression)
+const uploadObject = async (storageType: StorageType, key: string, value: string|Buffer) => {
+    if (storageType === StorageType.ObjectStorage) {
+        const params = {
+            Bucket: AWS_S3_BUCKET_NAME,
+            Key: key,
+            Body: compress(value),
+        };
+        return S3Client.putObject(params);
+    }
+
+    if (storageType === StorageType.FileStorage) {
+        // create directories if necessary
+        const keySegments = key.split('/');
+        for (let i = 1; i < keySegments.length; i++) {
+            const combinedSegments = keySegments.slice(0, i).join('/');
+            const realPath = path.resolve(`${__dirname}/../${combinedSegments}`);
```

Comment:
Folder structure changed, this will save it into the `/src` folder, add an extra `../`
```suggestion
            const realPath = path.resolve(`${__dirname}/../../${combinedSegments}`);
```